iroha/src/account.rs

```diff
@@ -5,70 +5,149 @@ use std::collections::BTreeMap;
+            name: String::from(vector[0]),
+            container: String::from(vector[1]),
+        }
+    }
+}
+
+impl Identifiable for Account {
+    type Id = Id;
+}
+
 pub mod isi {
     use super::*;
-    use crate::isi::Contract;
-    use iroha_derive::{IntoContract, Io};
-    use parity_scale_codec::{Decode, Encode};
+    use crate::isi::prelude::*;
+    use iroha_derive::*;
+    use std::ops::{Add, Sub};
+
+    #[derive(Clone, Debug, Io, Encode, Decode)]
+    pub enum AccountInstruction {
+        TransferAsset(
+            <Account as Identifiable>::Id,
+            <Account as Identifiable>::Id,
+            Asset,
+        ),
+    }
+
+    impl AccountInstruction {
+        pub fn execute(&self, world_state_view: &mut WorldStateView) -> Result<(), String> {
+            match self {
+                AccountInstruction::TransferAsset(
+                    source_account_id,
+                    destination_account_id,
+                    asset,
+                ) => Transfer::new(
+                    source_account_id.clone(),
+                    asset.clone(),
+                    destination_account_id.clone(),
+                )
+                .execute(world_state_view),
+            }
+        }
+    }
 
     /// The purpose of add signatory command is to add an identifier to the account. Such
     /// identifier is a public key of another device or a public key of another user.
-    #[derive(Clone, Debug, PartialEq, Io, IntoContract, Encode, Decode)]
-    pub struct AddSignatory {
-        pub account_id: Id,
-        pub public_key: [u8; 32],
-    }
+    impl Add<PublicKey> for Account {
+        type Output = Self;
 
-    /// The purpose of append role command is to promote an account to some created role in the
-    /// system, where a role is a set of permissions account has to perform an action (command or
-    /// query).
-    #[derive(Clone, Debug, PartialEq, Io, IntoContract, Encode, Decode)]
-    pub struct AppendRole {
-        pub account_id: Id,
-        pub role_name: String,
+        fn add(mut self, signatory: PublicKey) -> Self {
+            self.signatories.push(signatory);
+            self
+        }
     }
 
-    /// The purpose of create account command is to make entity in the system, capable of sending
-    /// transactions or queries, storing signatories, personal data and identifiers.
-    #[derive(Clone, Debug, PartialEq, Io, IntoContract, Encode, Decode)]
-    pub struct CreateAccount {
-        pub account_id: Id,
-        pub domain_name: String,
-        pub public_key: [u8; 32],
+    impl Sub<PublicKey> for Account {
+        type Output = Self;
+
+        fn sub(mut self, signatory: PublicKey) -> Self {
+            if let Some(index) = self.signatories.iter().position(|key| key == &signatory) {
+                self.signatories.remove(index);
+            }
+            self
+        }
     }
 
-    impl Instruction for CreateAccount {
+    impl Transfer<Account, Asset, Account> {
```

Comment:
I think it could be confusing that Transfer is here on account. When people think of transferring assets, they think of the asset have an action performed on it. Like someone handing over some cash. They don't think of themselves and the receiver as having a state change, because it is implicit and people are often blinded by the obviousness of it, almost like epistemic solipsism. However, this is quite a lot to ponder philosophically. When people ask what you can do with assets, you can tell them you can create, mint, demint, and transfer. Actions that are performed on an object should probably be with the object and not the object's owner.

Reply:
https://soramitsu.atlassian.net/browse/RH2-88